libraries/common/resizable_array.effekt

```diff
@@ -0,0 +1,234 @@
+/// shrink array when size / capacity falls below this threshold
+/// should be < 1/growFactor
+val shrinkThreshold = 0.4
+
+/// Number of elements in the dynamic array
+///
+/// O(1)
+def size[T](arr: ResizableArray[T]) = arr.rawSizePtr.get
+
+/// Allocate a new, empty dynamic array with given initial capacity
+def resizableArray[T](capacity: Int): ResizableArray[T] = {
+  ResizableArray(ref(0), ref(allocate(capacity)))
+}
+
+/// Allocate a new, empty dynamic array
+def resizableArray[T](): ResizableArray[T] = resizableArray(8)
+
+/// Throw an OutOfBounds exception if index is not a valid index into arr
+def boundsCheck[T](arr: ResizableArray[T], index: Int): Unit / Exception[OutOfBounds] = {
+  if (index < 0 && index >= arr.size) {
+    do raise(OutOfBounds(), "Array index out of bounds: " ++ show(index))
+  }
+}
+
+/// get the element at position index in the array
+///
+/// precondition: index is a valid index into the array
+///
+/// O(1)
+def unsafeGet[T](arr: ResizableArray[T], index: Int): T = {
+  arr.rawContentPtr.get.unsafeGet(index)
+}
+/// get the element at position index in the array
+///
+/// O(1)
+def get[T](arr: ResizableArray[T], index: Int): T / Exception[OutOfBounds] = {
+  arr.boundsCheck(index);
+  arr.unsafeGet(index)
+}
+
+/// set the element at position index in the array
+///
+/// precondition: index is a valid index into the array
+///
+/// O(1)
+def unsafeSet[T](arr: ResizableArray[T], index: Int, value: T): Unit = {
+  arr.rawContentPtr.get.unsafeSet(index, value)
+}
+
+/// set the element at position index in the array
+///
+/// O(1)
+def set[T](arr: ResizableArray[T], index: Int, value: T): Unit / Exception[OutOfBounds] = {
+  arr.boundsCheck(index);
+  arr.unsafeSet(index, value)
+}
+
+/// swap the elements at the given positions in the array
+///
+/// precondition: both are valid indices into the array
+///
+/// O(1)
+def unsafeSwap[T](arr: ResizableArray[T], index1: Int, index2: Int): Unit = {
+  val raw = arr.rawContentPtr.get
+  val tmp = raw.unsafeGet(index1)
+  raw.unsafeSet(index1, raw.unsafeGet(index2))
+  raw.unsafeSet(index2, tmp)
+}
+
+/// swap the elements at the given positions in the array
+///
+/// O(1)
+def swap[T](arr: ResizableArray[T], index1: Int, index2: Int): Unit / Exception[OutOfBounds] = {
+  arr.boundsCheck(max(index1, index2))
+  arr.unsafeSwap(index1, index2)
+}
+
+/// Change the dynamic to have exactly the given capacity
+///
+/// precondition: given capacity is at least the size of the array
+///
+/// O(n)
+def unsafeSetCapacity[T](arr: ResizableArray[T], capacity: Int): Unit = {
+  with on[OutOfBounds].panic
+  val oldRaw = arr.rawContentPtr.get
+  if (oldRaw.size != capacity) { 
+    val newRaw = array::allocate(capacity)
+    oldRaw.copy(0, newRaw, 0, arr.size)
+    arr.rawContentPtr.set(newRaw)
+  }
+}
+
+/// Change the dynamic to have exactly the given capacity
+///
+/// O(n)
+def setCapacity[T](arr: ResizableArray[T], capacity: Int): Unit / Exception[OutOfBounds] = {
```

Comment:
What about `resize`? If the type is called resizable array?

Reply:
I feel like `resize` might be interpreted as setting the size of the array (and filling with `0`s, e.g.), whereas this just resizes the backing array, not the data structure (it's almost a no-op when looking at it as an abstract data structure).

Reply:
We could implement `resize` (as in "make length k and fill"), too, of course.

Reply:
I am also fine with keeping as is. Maybe the reasoning can be explained a bit further in the comment to help users?